text/Issue 8 - Plug-in based architecture.md

```diff
@@ -0,0 +1,265 @@
+
+A crate `standardized` provides a library of standard `Feature`s to cover a number
+of uses:
+
+- garage door opened;
+- lock unlocked;
+- oven temperature;
+- ...
+
+Use of a `Feature` is decided per-channel, by the Adapter developer.
+
+Third-party developers are encouraged to pick a standardized `Feature`, but
+are not forced to. Additional instances of `Feature` can easily be published
+by third-party developers to third-party crates.
+
+### Implementation
+
+
+```rust
+struct Feature<T> where T: Message {
+  /// Keys used to discover this feature.
+  ///
+  /// For instance: `["x-oven-temperature", "oven-temperature"]`, where
+  /// `x-oven-temperature` was the name used prior to standardization and
+  /// `oven-temperature` is the standardized name.
+  pub keys: Vec<String>,
+
+  phantom: PhantomData<T>
+}
+```
+
+
+## Propagating the changes
+
+We modify the definition of channels to use `Feature<T>` instead of
+`ChannelKind`. Otherwise, this definition is unchanged. We similarly modify
+`add_getter`/`add_setter` to use this new definition of `Channel<T>`.
+
+In the implementation of `Adapter`, type `Value` is replaced by `Any`.
+
+In the Taxonomy API, type `Value` is replaced by `JSON`, with a companion `BinaryParts`.
+
+### Implementation
+
+```rust
+pub enum WatchEvent {
+  EnterRange {
+    from: Id<Getter>,
+    json: JSON,
+    binary: BinaryParts
+  },
+  ExitRange {
+    from: Id<Getter>,
+    json: JSON,
+    binary: BinaryParts
+  },
+  // ...
+}
+
+trait API {
+  fn fetch_values(&self, Vec<GetterSelector>, user: User) ->
+    (ResultMap<Id<Getter>, Option<JSON>, Error>, BinaryParts);
+  fn send_values(&self, TargetMap<SetterSelector, JSON>, binary: &BinaryParts,
+    user: User) -> ResultMap<Id<Setter>, (), Error>;
+}
+```
+
+```rust
+trait Adapter {
+  fn fetch_values(&self, set: &[Id<Getter>], user: User) ->
+    ResultMap<Id<Getter>, Option<Box<Any>>, Error>;
+
+  fn send_values(&self, values: HashMap<Id<Setter>, &Any>, user: User) ->
+    ResultMap<Id<Setter>, (), Error>;
+}
+```
+
+
+# Consequences
+
+## Conflicts
+
+Having the same `Message` in several Adapters causes no conflict.
+
+Several instances of `Feature` may share some or all `keys`, possibly with
+different instances of `Message`. This will commonly be the case for features that
+have not been standardized yet but that are implemented by several
+devices/adapters, or that have been standardized recently. This is considered
+normal. The AdapterManager will handle this, by dispatching per key, rather
+than by `Feature`.
+
+If several Adapters use distinct instances of `Message` for the same `keys`, or
+use the `RawBinary`/`RawJSON` `Message`, this may cause a request to the Foxbox to
+be fail due to a parse error or a serialization error with some adapters and
+succeed for others. Similarly, this may cause a request to the Foxbox to return
+messages with completely unrelated serialization formats for several adapters
+implementing the same key.
```

Comment:
# Getters

Adapter 1 offers a Getter for `x-heater-temperature`, using the `foxbox_standardized::message::Temperature` for (de)serialization. This `Message` creates temperatures with either format `{"C": 27}` or `{"F": 100}`.

Adapter 2 offers a Getter for `x-heater-temperature`, using a custom `Message` that creates temperatures with format `{"unit": "F", "value": 100}`.

Adapter 3 forgets entirely to use a unit and returns `100`.

Web application places a fetch request for all temperatures. The result is

```json
[{
  "heater_1": {"C": 100},
  "heater_2": {"unit": "C", "value": 100"},
  "heater_3": 100
}]
```

Which is not very convenient. Note, by the way, that Adapter 3 can even cause bodily harm, because there is no unit involved.

# Setters

With the same adapters, attempting to set the temperature of all `x-heater-temperature` channels to `{"C": 19}` will return the following result

```json
[{
  "heater_1": Ok,
  "heater_2": Error("ParseError: ...")
  "heater_3": Error("ParseError: ...")
}]
```

Which is not very nice, either.